engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
-	if s.crash {
-		s.log.Fatal().Msgf(
-			"Crashing as finalization reached requested "+
-				"stop height %d and the highest executed block is (%d - 1)",
-			s.stopHeight,
-			s.stopHeight,
-		)
+	log := s.log.With().
+		Stringer("requested_stop", s.stopBoundary).
+		Uint64("last_executed_height", s.stopBoundary.StopHeight).
+		Stringer("last_executed_id", s.stopBoundary.stopAfterExecuting).
+		Logger()
+
+	s.stopped = true
+	log.Warn().Msg("Stopping as finalization reached requested stop")
+
+	if s.stopBoundary != nil && s.stopBoundary.ShouldCrash {
+		// TODO: crash more gracefully or at least in a more explicit way
+		log.Fatal().Msg("Crashing as finalization reached requested stop")
 		return
 	}
+}
 
-	s.log.Debug().
-		Int8("previous_state", int8(s.state)).
-		Int8("new_state", int8(StopControlPaused)).
-		Msg("StopControl state transition")
+func (s *StopControl) handleVersionBeacon(
+	height uint64,
+) error {
+	if s.nodeVersion == nil {
+		return nil
+	}
 
-	s.state = StopControlPaused
+	if s.versionBeacon != nil && s.versionBeacon.SealHeight >= height {
+		// we already processed this or a higher version beacon
+		return nil
+	}
 
-	s.log.Warn().Msgf(
-		"Pausing execution as finalization reached "+
-			"the requested stop height %d",
-		s.stopHeight,
-	)
+	vb, err := s.versionBeacons.Highest(height)
+	if err != nil {
+		return fmt.Errorf("failed to get highest "+
+			"version beacon for stop control: %w", err)
+	}
 
-}
+	if vb == nil {
+		// no version beacon found
+		// this is unexpected as there should always be at least the
+		// starting version beacon, but not fatal.
+		// It can happen if the node starts before bootstrap is finished.
+		s.log.Info().
+			Uint64("height", height).
+			Msg("No version beacon found for stop control")
+		return nil
+	}
 
-// executingBlockHeight should be called while execution of height starts,
-// used for internal tracking of the minimum possible value of stopHeight
-func (s *StopControl) executingBlockHeight(height uint64) {
-	// TODO: should we lock here?
+	s.log.Info().
+		Uint64("vb_seal_height", vb.SealHeight).
+		Uint64("vb_sequence", vb.Sequence).
+		Msg("New version beacon found")
 
-	if s.state == StopControlPaused {
-		return
+	// this is now the last handled version beacon
+	s.versionBeacon = vb
+
+	// this is a new version beacon check what boundary it sets
+	stopHeight, shouldBeSet, err := s.getVersionBeaconStopHeight(vb)
+	if err != nil {
+		return err
+	}
+
+	set := s.stopBoundary != nil
+
+	if !set && !shouldBeSet {
+		// all good, no stop boundary set
+		return nil
+	}
+
+	var newStop *stopBoundary
+	if shouldBeSet {
+		newStop = &stopBoundary{
+			StopParameters: StopParameters{
+				StopHeight:  stopHeight,
+				ShouldCrash: s.crashOnVersionBoundaryReached,
+			},
+			fromVersionBeacon: true,
+		}
+	}
+
+	err = s.unsafeSetStopParameters(newStop, true)
+	if err != nil {
+		s.log.Warn().
+			Err(err).
+			Msg("Cannot change stop boundary when detecting new version beacon")
```

Comment:
Could we instead log within `unsafeSetStopParameters` and remove the error return here?

Reply:
I reduced the log to info, because this can happen for a few reasons described on `canChangeStop`. I still want to keep the error due to `unsafeSetStopParameters` being used in 2 places, one here and one in the admin command and the error is convenient for the admin command.
[b61af2a](https://github.com/onflow/flow-go/pull/4342/commits/b61af2a8773709c672ee17ccd6e777b08aa20daa)

I will address the error handling.